### ISSUE TITLE
Train time seems to depend heavily on image count

### ISSUE THREAD

**Issue by StephenChan** [coralnet/pyspacer#73]
Some recent train times:

- 1052 (VGG16) completed in 52 hours: 56,000 images x 30 points
- 1656 (VGG16) completed in 50 hours: 65,000 images x 50 points
- 295 (VGG16) completed in 38 hours: 63,000 images x 10 points
- 1374 (VGG16) completed in 21 hours: 34,000 images x 30 points
- 1288 (EffNet) completed in 7 hours: 29,000 images x 10 points
- 3182 (EffNet) completed in 1.6 hours: 3800 images x 100 points
- 3342 (EffNet) completed in 1.5 hours: 1200 images x 1000 points

There are a few more train times in issue #53 too.

Looking at 1052 vs. 1656, obviously there are just some fluctuations on how fast training jobs run, but overall the points per image doesn't seem like as big of a factor as one might expect. Especially looking at the EfficientNet times which we should be more interested in going forward. It's definitely got some effect, but not close to linear. On the other hand, the image count has a closer to linear influence than we might expect, given that the number of training batches is determined by the number of points (features), and given that there's no parallelization in the whole training process.

And there are no anomalies with the feature vector filesize; that's pretty much linearly related to point count, at least for EfficientNet:

- EffNet 10 points (1288) examples: i3653915 = 56 KB, i1108305 = 56 KB
- EffNet 100 points (3182) examples: i3655456 = 546 KB, i3655386 = 546 KB
- EffNet 1000 points (3342) examples: i3279111 = 5.3 MB, i3278833 = 5.3 MB

As for what timescale we're talking about per image:

- 1052: 52 hours / 56,000 images = 3.3 seconds/image
- 1656: 50 hours / 65,000 images = 2.8 seconds/image
- 295: 38 hours / 63,000 images = 2.2 seconds/image
- 1374: 21 hours / 34,000 images = 2.2 seconds/image
- 1288: 7 hours / 29,000 images = 0.9 seconds/image
- 3182: 1.6 hours / 3800 images = 1.5 seconds/image
- 3342: 1.5 hours / 1200 images = 4.5 seconds/image

Keep in mind that we always run 10 training epochs, so the seconds per image per epoch would be 0.33, 0.28, etc.

Here's the first thing I'm thinking of looking into, from issue #53:

> Try to reuse S3 connections as long as possible instead of using one connection per feature vector download

On every feature vector download (which happens per image per epoch), we currently call `config.get_s3_conn()` which gets a new boto3 resource using the AWS credentials:

```python
return boto3.resource(
    's3',
    region_name=AWS_REGION,
    aws_access_key_id=AWS_ACCESS_KEY_ID,
    aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
)
```

And a connection/authentication step like this might be a significant factor at a timescale of 0.33s, 0.09s, etc.

I'm not sure if the resource expires after a certain point. So that's a detail to look out for, but most likely not limiting enough to kill our hope of a speedup (on its own).

**Comment by StephenChan** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [support for large training jobs](https://trello.com/c/gbO1Bu5d/18-support-for-large-training-jobs)

**Comment by StephenChan** [comment]
Actually, from the [blog post](https://coralnet.ucsd.edu/blog/a-new-deep-learning-engine-for-coralnet/), on point 2:

> Classifier Online Training Time
>
> While the Multi-layer Perceptron classifier is more accurate than than the Logistic Regression classifier, we wanted to ensure that it wouldn’t be prohibitively slow or expensive. ... As shown in the plot below, the training speed is nearly the same as Logistic Regression classifier on small sources and even faster on large sources.

I think the bar plot in that section supports that statement. So unless there's some other reason why LR would be more desirable than MLP for smaller sources, I guess the logic was "all else being equal, it's simpler to say MLP = EfficientNet, LR = VGG16, rather than mixing and matching".

**Comment by kriegman** [comment]
So, this looks like MLP is 2 layer for less than 50k points and 3 layer for
more than 50k points.
3 layer meaning 2 hidden layers + output layer which has the number of
outputs = number of classes.

D

**Comment by beijbom** [comment]
In my experience, one can turn down the tolerance parameter ('tol
<https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html>')
A LOT without affecting the accuracy. It defaults to 10^-4, but I have used
10^-1 recently. This makes the training converge much faster.

Ofcourse, the responsible thing would be to sanity check this on some real
coralnet training data before changing it. But I thought I'd at least offer
the suggestion. You could probably go to say 10^-3 or 10^-2 without doing
any testing.

But yeah, looking at the measurements, the first thing to do is to kindly
ask folks to move over to EfficientNet. That one-time feature compute cost
will be made up quickly by shorter train times.

**Comment by StephenChan** [comment]
Thanks for the tip on the tolerance parameter! For the extractor choice on CoralNet, we should definitely point out that EfficientNet results in faster runtimes if we aren't already.

**Comment by kriegman** [comment]
And more than that, greater accuracy 9.5 times out of 10

**Comment by kriegman** [comment]
Actually. Maybe we should be more draconian and just not retrain for any
vgg sources and only do inference. And let users know that if they want
improved classifiers, they need to upgrade to efficient net.

Also I’m not sure if we give an option for vgg for new sources, but we
could remove the option.

David

**Comment by StephenChan** [comment]
I'm warming up to the idea of pushing VGG out more proactively, but also hesitant to do it when we have a fair number of other things going on. (As long as it's not the cause of urgent issues like site unresponsiveness, but I still don't think it is)

Though maybe a more moderate option would be to say, VGG is no longer supported for training/retraining when above a certain number of images/annotations.

**Comment by kriegman** [comment]
Aiming for simplicity. Rather than doing anything to make the vgg
experience better, the effort should be toward deprecating or just ignoring…

Eg, no readi. For fast LR training if it’s only for vgg. Faster MLP is more
useful
David

**Comment by StephenChan** [comment]
I'll keep that in mind - I think most/all of the ideas tossed around recently would benefit both or just benefit MLP.

**Comment by StephenChan** [comment]
Once PR #80 or equivalent is merged, think we'll have appreciable enough performance improvements to close this issue; stacking a x6 speedup from that and x2 from PR #53. Train time will still depend on image count, but not to the extent it did before.

**Comment by StephenChan** [comment]
> Once PR https://github.com/coralnet/pyspacer/pull/80 or equivalent is merged, think we'll have appreciable enough performance improvements to close this issue

And now that PR is merged.
